### ISSUE TITLE
Certain seeds cause dungeon generation to get stuck in an infinite loop

### ISSUE THREAD

**Issue by kpyrkosz** [diasurgical/devilutionX#5346]
### Operating System

Linux x64

### DevilutionX version

1.4.1 (latest release)

### Describe

Some seeds, such as the 3 in the attached test case cause the program to hang during dungeon generation. I've attached a test case demonstrating the issue.

### To Reproduce

```
TEST(Drlg_l2, seeds_stuck_in_an_infinite_loop)
{
        leveltype = GetLevelType(5);
        int bad_seeds[3] = { 606, 5248, 1054405 };

        for (int i = 0; i < 3; ++i) {
                pMegaTiles = std::make_unique<MegaTile[]>(GetTileCount(leveltype));
                CreateDungeon(bad_seeds[i], ENTRY_MAIN);
        }
}
```

### Expected Behavior

Program does not hang

**Comment by StephenCWills** [comment]
As was mentioned in Discord, you can also reproduce using debug commands in a Multiplayer game.
```
restart 5 606
changelevel 5
```

And it was confirmed to hang in vanilla Diablo as well.

**Comment by AJenbo** [comment]
It's stuck moving left and right between 31x11 and 38x11.

So looks like it decided to plow though the C corner and wall of the room, and now is stuck between the map edge and B corner of the room.

**Comment by MohitS05** [comment]
I suppose a solution can be to add a std::future call and add a timeout, incase we don't get generation within time we stop that thread and try to regenerate with a different seed. I am new to this codebase, if pointed to generation code I'll be happy to implement it.

**Comment by StephenCWills** [comment]
If you intend to time it out, you can avoid the thread management and use `SDL_GetTicks()`. However, a better solution might be to detect the conditions for the infinite loop and bail. I don't know if that's feasible.

**Comment by AJenbo** [comment]
I think it's doable

**Comment by alex-polosky** [comment]
> ... However, a better solution might be to detect the conditions for the infinite loop and bail. I don't know if that's feasible.

Isn't this the [halting problem](https://en.wikipedia.org/wiki/Halting_problem)?

**Comment by StephenCWills** [comment]
No, a key component of the halting problem is highlighted in that first sentence on Wikipedia.

> In computability theory, the halting problem is the problem of determining, from a description of **an arbitrary computer program** and an input, whether the program will finish running, or continue to run forever.

We would only need to determine what the conditions are for this particular loop to become infinite, not any arbitrary algorithm. Then we could hardcode an edge case or workaround for it.

**Comment by NiteKat** [comment]
[single_0 (4).zip](https://github.com/user-attachments/files/30965697/single_0.4.zip)
Attached is a save file that hangs when trying to load floor 5. Not sure if it is the exact bug originally reported here, but it matches the description of this issue so adding it as an example.
